### ISSUE TITLE
Should `authorize_response()` check the CSRF state token?

### ISSUE THREAD

**Issue by hurrymaplelad** [lepture/flask-oauthlib#206]
I'd like to auth with Google and protect my oauth2 callback from CSRF using the state token, but I'm not sure how I should validate the state querystring arg returned by Google to my callback.

I can do it manually with something like:
```py
def oauth2callback():
    forgery_token = session.get('ga_state', None)
    if not forgery_token or forgery_token != request.args.get('state', None):
        return make_response('Invalid state parameter.', 401)
     del session['ga_state']
    resp = google.authorized_response()
    # ...
```

But it seems oauthlib [has some plumbing for this](https://github.com/idan/oauthlib/blob/62ee76f22e8b0a53b2eda4a88e2891ed6578083b/oauthlib/oauth2/rfc6749/parameters.py#L179) in `parameters.parse_authorization_code_response`.

Any thoughts on passing state in to `authorized_response()` and plumbing it down to either [`parse_authorization_code_response`](https://github.com/idan/oauthlib/blob/62ee76f22e8b0a53b2eda4a88e2891ed6578083b/oauthlib/oauth2/rfc6749/parameters.py#L179) or through `prepare_request_body` assuming [oauthlib calls `BaseClient.prepare_token_request` instead of `parameters.prepare_token_request`](https://github.com/idan/oauthlib/issues/344)?

Seems strange to me that LinkedIn example [passes a state token](https://github.com/lepture/flask-oauthlib/blob/master/example/linkedin.py#L16) but [doesn't appear to do any of the above checks in the callback](https://github.com/lepture/flask-oauthlib/blob/master/example/linkedin.py#L47).  Am I missing something?

**Comment by emilecaron** [comment]
@hurrymaplelad I think it should.
Still it's not that much effort to do it it yourself:
`state` is accessible in `request.args['state']` in the authorize method.

**Comment by hurrymaplelad** [comment]
@emilecaron thanks!  Agreed, not much effort to add the check.  

The bigger issue to me is it took a while to determine whether I needed to add the check or the library was already doing it.  The [examples that pass state](https://github.com/lepture/flask-oauthlib/blob/master/example/linkedin.py#L16) don't check it, suggesting it's handled by the library.  I suspect some folks think they have forgery protection that don't.

**Comment by lepture** [comment]
@hurrymaplelad I think the library should check the `state`

**Comment by emilecaron** [comment]
@lepture  It's not that easy. Problem is storing the thing until google sends its authorize request. Using the session like in the examples is kind of not cool, it breaks when multiple users try to auth at the same time. There are good solutions, but it depends a lot on what you're trying to do.

**Comment by ColeKettler** [comment]
Yeah, I just ran into this, as well. The lack of automatic CSRF protection should be made more apparent, I didn't realize that nothing was happening until I read the source.

I think that verifying the integrity of the state is a good idea, since the OAuth 2.0 spec [requires](https://tools.ietf.org/html/rfc6749#section-4.1.27) that the authorization server return the unmodified `state` parameter in the authorization response.

As far as storage goes, perhaps a decorator to register a function to consistently generate the state, along the lines of `tokengetter`? That does put the burden on the user to implement it, but it allows flexibility as far as generation and storage go. Then I could do something like:

    @remoteapp.stategetter
    def get_state():
        return generate_csrf_token()
